app/collectors/launchTemplates.scala

```diff
@@ -0,0 +1,122 @@
+package collectors
+
+import java.time.Instant
+import agent._
+import conf.AWS
+import controllers.routes
+import play.api.mvc.Call
+import software.amazon.awssdk.services.autoscaling.AutoScalingClient
+import software.amazon.awssdk.services.ec2.Ec2Client
+import software.amazon.awssdk.services.ec2.model.{
+  DescribeLaunchTemplateVersionsRequest,
+  LaunchTemplateVersion => AwsLaunchTemplateVersion
+}
+import utils.Logging
+
+import scala.jdk.CollectionConverters._
+import scala.language.postfixOps
+import scala.util.Try
+
+class LaunchTemplateCollectorSet(accounts: Accounts)
+    extends CollectorSet[LaunchTemplateVersion](
+      ResourceType("launch-templates"),
+      accounts,
+      Some(Regional)
+    ) {
+  val lookupCollector
+      : PartialFunction[Origin, Collector[LaunchTemplateVersion]] = {
+    case amazon: AmazonOrigin =>
+      AWSLaunchTemplateCollector(
+        amazon,
+        resource,
+        amazon.crawlRate(resource.name)
+      )
+  }
+}
+
+case class AsgLaunchTemplate(id: String, version: String)
+
+case class AWSLaunchTemplateCollector(
+    origin: AmazonOrigin,
+    resource: ResourceType,
+    crawlRate: CrawlRate
+) extends Collector[LaunchTemplateVersion]
+    with Logging {
+
+  val asgClient: AutoScalingClient = AutoScalingClient.builder
+    .credentialsProvider(origin.credentials.provider)
+    .region(origin.awsRegionV2)
+    .overrideConfiguration(AWS.clientConfig)
+    .build
+
+  val ec2Client: Ec2Client = Ec2Client.builder
+    .credentialsProvider(origin.credentials.provider)
+    .region(origin.awsRegionV2)
+    .overrideConfiguration(AWS.clientConfig)
+    .build
+
+  def crawl: Iterable[LaunchTemplateVersion] = {
+    val asgs =
+      asgClient.describeAutoScalingGroups().autoScalingGroups().asScala.toList
+    val launchTemplates = asgs
+      // not all asgs have launch templates (they might have launch configurations instead
+      .filter(asg => asg.launchTemplate() != null)
+      .map(asg =>
+        AsgLaunchTemplate(
+          asg.launchTemplate().launchTemplateId(),
+          asg.launchTemplate().version()
+        )
+      )
+
+    launchTemplates.map { lt =>
+      // might need to filter out asgs which don't have launch templates?
+      val requestLt = DescribeLaunchTemplateVersionsRequest.builder
+        .launchTemplateId(lt.id)
+        .versions(lt.version)
+        .build
+      val template = ec2Client
+        .describeLaunchTemplateVersions(requestLt)
+        .launchTemplateVersions()
+        .asScala
+        .head
```

Comment:
Are we sure we don't need to capture the other versions? I think it's rare to do versioned deployments at present, but I'm hoping that will start to change soon, maybe even this quarter! (Versioned deployments would unlock automated rollback of failed deplyments, for example)

Since we're motivated by deletinbg old AMIs in Amigo, this would create a decision about how far back we'd like our launch templates to still be launchable, but that's a decision for Amigo to make, rather than Prism. That suggests Prism should be exposing at least as much information as Amigo needs to make that decision, which would certainly more than just the most recent?

I may have misunderstood this bit, in which case apologies!

Reply:
Hi again, I refactored this in https://github.com/guardian/prism/pull/918/commits/9392471a8d4590162cb9becde524f4f5d87d919f to fetch all versions of all launch templates. Unfortunately the deploy failed, I think because several of the launch template collectors repeatedely failed, and prism only passes the healthcheck once it has managed to fetch everything for each account at least once (see https://logs.gutools.co.uk/s/devx/goto/6c624950-44fd-11ef-8041-ff280ea02022)

Reply:
I'm not sure where to go next, but I am running out of time to work on this, so am tempted to suggest we stick with the original solution, unless you have any better ideas? For the deploy tools account alone there are 581 launch templates, so my guess is that there are quite a lot more for accounts like frontend. Prism should be able to handle this, but just means thinking a lot more about paging etc than we have had to before for this application.

Reply:
Hey @adamnfish  - what do you think we should do here?